reconcile/utils/saasherder.py

```diff
@@ -344,6 +364,62 @@ def _validate_image_tag_not_equals_ref(
             )
             self.valid = False
 
+    @staticmethod
+    def _get_upstream_jobs(
+            jjb: JJB,
+            all_jobs: Mapping[str, dict],
+            url: str,
+            ref: str,
+    ) -> Iterable[UpstreamJob]:
+        results = []
+        for instance, jobs in all_jobs.items():
+            for job in jobs:
+                job_repo_url = jjb.get_repo_url(job)
+                if url != job_repo_url:
+                    continue
+                job_ref = jjb.get_ref(job)
+                if ref != job_ref:
+                    continue
+                results.append(UpstreamJob(instance, job['name']))
+        return results
+
+    def validate_upstream_jobs(
+            self,
+            jjb: JJB,
+    ):
+        all_jobs = jjb.get_all_jobs(job_types=["build"])
+        pattern = r'^[0-9a-f]{40}$'
+        for sf, rt, t in self:
+            sf_name = sf['name']
+            rt_name = rt['name']
+            url = rt['url']
+            ref = t['ref']
+            if re.search(pattern, ref):
+                continue
+            upstream = t.get("upstream")
+            if upstream:
+                if isinstance(upstream, str):
+                    # skip v1 saas files
+                    continue
+                upstream_job = UpstreamJob(upstream['instance']['name'], upstream['name'])
+                possible_upstream_jobs = self._get_upstream_jobs(jjb, all_jobs, url, ref)
+                found_jobs = [j for j in all_jobs[upstream_job.instance] if j["name"] == upstream_job.job]
+                if found_jobs:
+                    if upstream_job not in possible_upstream_jobs:
```

Comment:
to verify i understand this condition: this is true if the upstream job exists by name but the url and/or ref do not match - so basically a configuration error.

Reply:
correct